server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java

```diff
@@ -105,8 +110,23 @@ private void updateRetentionPolicy() throws IOException {
         translogDeletionPolicy.setTranslogGenerationOfLastCommit(lastGen);
         translogDeletionPolicy.setMinTranslogGenerationForRecovery(minRequiredGen);
 
-        softDeletesPolicy.setLocalCheckpointOfSafeCommit(
-            Long.parseLong(safeCommit.getUserData().get(SequenceNumbers.LOCAL_CHECKPOINT_KEY)));
+        final long localCheckpointOfSafeCommit = Long.parseLong(safeCommit.getUserData().get(SequenceNumbers.LOCAL_CHECKPOINT_KEY));
+        softDeletesPolicy.setLocalCheckpointOfSafeCommit(localCheckpointOfSafeCommit);
+
+        final long docCountOfSafeCommit = getDocCountOfSafeCommit();
```

Comment:
Drive by comment: I think this will do IO while holding the monitor. This would then prevent concurrent `acquireIndexCommit` and `releaseIndexCommit` calls. A quick scrape of usages did not find any where this would hurt, but I would anyway suggest to do the calculation of `minimumReasonableRetainedSeqNo` outside the monitor instead.

Reply:
Yes, it will. We're already in a method that `throws IOException` but you're right, it doesn't look like it really does do much IO as it currently stands. Moved outside the mutex in 5860c41.